### PR TITLE
Fixed einsum compatibility/performance issues

### LINKED ISSUE #32591: `einsum` is ~20X slower than manually multiplying and summing
## 🐛 Bug
A manual multiplication and summation `(a * b).sum(dim = (-3, -2, -1))` is about 20X faster than the equivalent `einsum`.

## To Reproduce

```Python3
a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device = "cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device = "cuda")

# Warmup
for i in range(100): output1 = (a * b).sum(dim = (-3, -2, -1))

# Method 1
with torch.autograd.profiler.profile(use_cuda = True) as prof:
  output1 = (a * b).sum(dim = (-3, -2, -1))
print(prof.key_averages().table(sort_by="cuda_time_total"))

# Warmup
for i in range(100):
  (a2, b2) = torch.broadcast_tensors(a, b)
  output2 = torch.einsum("...ijk, ...ijk -> ...", a2, b2)

# Method 2
with torch.autograd.profiler.profile(use_cuda = True) as prof:
  (a2, b2) = torch.broadcast_tensors(a, b)
  output2 = torch.einsum("...ijk, ...ijk -> ...", a2, b2)
print(prof.key_averages().table(sort_by="cuda_time_total"))
```
```
-------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
-------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
sum      28.93%           43.848us         28.93%           43.848us         43.848us         65.81%           8.978ms          8.978ms          1                
mul      71.07%           107.713us        71.07%           107.713us        107.713us        34.19%           4.665ms          4.665ms          1                
-------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 151.561us
CUDA time total: 13.644ms

---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                   Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
einsum                 4.05%            59.449us         97.88%           1.435ms          1.435ms          48.62%           123.798ms        123.798ms        1                
bmm                    85.15%           1.248ms          85.15%           1.248ms          1.248ms          45.83%           116.694ms        116.694ms        1                
reshape                1.35%            19.770us         6.36%            93.237us         46.619us         2.76%            7.031ms          3.516ms          2                
clone                  3.96%            58.021us         3.96%            58.021us         29.011us         2.75%            7.012ms          3.506ms          2                
broadcast_tensors      0.79%            11.624us         2.12%            31.095us         31.095us         0.01%            31.008us         31.008us         1                
expand                 1.33%            19.471us         1.33%            19.471us         9.736us          0.01%            20.480us         10.240us         2                
permute                1.54%            22.535us         1.54%            22.535us         4.507us          0.01%            19.327us         3.865us          5                
view                   0.78%            11.442us         0.78%            11.442us         5.721us          0.00%            4.094us          2.047us          2                
_unsafe_view           1.05%            15.446us         1.05%            15.446us         7.723us          0.00%            3.968us          1.984us          2                
---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 1.466ms
CUDA time total: 254.614ms
```

## Expected behavior

`einsum` to be no slower than manual. I actually expected / hoped for it to be faster as it avoids creating the large intermediary result of the multiplication.

## Environment

The environment is Google Colab with a fresh GPU instance on which `!pip install -q torch==1.4.0 torchvision==0.5.0` has been run as Colab is still on `1.3.1` by default.

```
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.17.5
[pip3] torch==1.4.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.5.0
[conda] Could not collect
```

cc @ngimel @vincentqb @vishwakftw @jianyuh @nikitaved @pearu @VitalyFedyunin @mruberry

### LINKED ISSUE #41467: einsum running 100x slower than expected when controlling for number of ops
## 🐛 Bug

For two simple einsum cases, Case 2 is running ~30% slower than Case 1, despite needing to perform 101x fewer ops.

## To Reproduce

**Case 1**
```
a = torch.randn(10000, 100, 101).cuda()
b = torch.randn(10000, 101, 3).cuda()

%%time
for _ in range(10000):
    torch.einsum("bij, bjf->bif", a, b)
    torch.cuda.synchronize()
```
`CPU times: user 26.7 s, sys: 18.8 s, total: 45.5 s
Wall time: 45.5 s`

**Case 2**
```
c = torch.randn(10000, 100, 1).cuda()
d = torch.randn(10000, 100, 1, 3).cuda()

%%time
for _ in range(1000):
    torch.einsum("bic, bicf->bif", c, d)
    torch.cuda.synchronize()
```
`CPU times: user 37.7 s, sys: 23.7 s, total: 1min 1s
Wall time: 1min 1s`

## Expected behavior

I believe the number of ops needed for each case should be approximately:
**Case 1:** 10,000 x 100 x 101 x 3 ( x 2)
**Case 2:** 10,000 x 100 x 1 x 3 (x 2) 
(The factor of 2 is to account for the multiply-accumulate operation used in matrix multiplication)

Despite requiring 101x fewer ops and operating over smaller tensors than Case 1, in the example above Case 2 has a wall clock time 34% greater than Case 1.
Are there methods/kernels that can be used which are more optimised towards the later case?

## Environment

Performed using a fresh Google Colab instance.

PyTorch version: 1.5.1+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.5.1+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.6.1+cu101
[conda] Could not collect

## Additional context
Using `opt_einsum==3.2.1` did not improve performance on Case 2.
```
from opt_einsum import contract
c = torch.randn(10000, 100, 1).cuda()
d = torch.randn(10000, 100, 1, 3).cuda()

%%time
for _ in range(10000):
    contract("bic, bicf->bif", c, d)
    torch.cuda.synchronize()
```
`CPU times: user 39 s, sys: 23.7 s, total: 1min 2s
Wall time: 1min 2s`

cc @ngimel @vincentqb @vishwakftw @jianyuh @nikitaved @pearu @VitalyFedyunin

### LINKED ISSUE #45854: torch.einsum ellipsis broadcasting is too strict and differs from numpy
## 🐛 Bug

torch.einsum requires ellipsis to represent a fixed number of dimensions across all operands while numpy is more flexible for broadcasting.

## To Reproduce

```
>>> np.einsum('i...k,k...j->i...j', torch.rand(2, 1, 3), torch.rand(3, 2)).shape
(2, 1, 2)

>>> torch.einsum('i...k,k...j->i...j', torch.rand(2, 1, 3), torch.rand(3, 2)).shape
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/scratch/heitorschueroff/pytorch/torch/functional.py", line 344, in einsum
    return _VF.einsum(equation, operands)  # type: ignore
RuntimeError: ellipsis must represent 1 dimensions in all terms
```

## Expected behavior

torch.einsum should match numpy.einsum behavior

## Environment

PyTorch version: 1.8.0a0+3ab88c3
Is debug build: True
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.105
GPU models and configuration: 
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 418.116.00
cuDNN version: Could not collect
HIP runtime version: N/A

cc @mruberry @rgommers @vincentqb @vishwakftw @jianyuh @nikitaved @pearu

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46398 Fixed einsum compatibility/performance issues**

#### Update: Fixed BC breaking change after reverting

This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option as requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

Differential Revision: [D24860367](https://our.internmc.facebook.com/intern/diff/D24860367)

### CODE REVIEW

File: aten/src/ATen/native/Linear.cpp
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:143]: Maybe file and cite a follow-up performance issue for an investigation or TODO into the possible performance improvements here?

File: aten/src/ATen/native/Linear.cpp
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:153]: const auto

File: aten/src/ATen/native/Linear.cpp
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:161]: Technically `i` would be `std::string::size_type`. You can get around the type mismatch by writing

```
const auto lhs_length = lhs.length()
for (auto i = decltype(lhs_length){0}; i < lhs_length; ++i) { 
```

If you really wanted to be fancy you could even use functions like `std::result_of` to avoid explicitly creating `lhs_length`, but I think it makes the code more readable.

File: aten/src/ATen/native/Linear.cpp
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:187]: Separate the increment to curr_op from the TORCH_CHECK to make this more readable.

File: test/test_linalg.py
Comment (mruberry) [inline test/test_linalg.py:928]: I think this check is redundant with the tensors being equal below?

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:269]: "left-hand_side" of the equation

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:270]: "with each operand separated by a comma (","). "

Comment (mruberry) [reply torch/functional.py:270]: "right-hand-side" of the equation, starting after an arrow ("->"), specifies

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:271]: "These must appear in at least one operand." -> "At least one operand must appear."

On the lhs, rhs, or both?

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:272]: The arrow ("->") may be omitted, in which case the output is

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:273]: Provide a brief parenthetical like: (i.e. "ijk" is equivalent to ...)

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:268]: It's wild how much documentation was packed into this argument description, and how little is in the function description itself. Maybe we can move some of this there and improve the description with examples?

File: aten/src/ATen/native/Linear.cpp
Comment (heitorschueroff) [inline aten/src/ATen/native/Linear.cpp:226]: Add comment explaining the +1 and why we need the ell_num_dim.

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:266]: An example here would probably be helpful.

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:265]: I'm not sure what a "multilinear expression" is. Also, are these equations really written using "Einstein notation?" My understand of Einstein notation isn't perfect but that description seems like a stretch. The [Wikipedia article](https://en.wikipedia.org/wiki/Einstein_notation) doesn't seem to have a 1:1 match with this equation, either. What do you think?

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:269]: "two modes" makes it sound like there's a `mode` parameter.

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:277]: What's actually computed still isn't clear at this point. Is there a mathematical formula that can be written before this? Then the notation could relate to that formula.

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:284]: What does it mean to "sum over" a dimension or dimensions?

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:281]: I'm not sure what "dimensions under the subscript" means

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:286]: Is it just that the dimensions can be reordered in the output? How should a user understand the "default" ordering/naming of output dimensions?

File: torch/functional.py
Comment (heitorschueroff) [reply torch/functional.py:265]: You're right, It's based on Einstein notation but not the same and to be honest I'm not entirely sure what a multilinear expression is either, I guess it's a system of linear equations with multiple lines? haha. Just an idea but, maybe we should get someone more familiar with the math behind each operator to review docs besides just the engineer.

Comment (mruberry) [reply torch/functional.py:265]: Maybe, but the typical reader won't know these terms, either. The documentation should break down what the operation does in terms that any reader can understand (possibly after consulting Wikipedia).

**Comment by JackCaoG** [comment]
Hi @heitorschueroff , xla side fix is ready. I will merge it once this change merged to pytorch.

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:266]: "based on the Einstein summation convention"

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:268]: no comma needed here

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:269]: "the Einstein summation convention" to be consistent with the above

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:270]: "described below"

**Review by mruberry — APPROVED** [review]
Docs look a lot better and the benchmarks look good. Made a few small grammar comments.

File: aten/src/ATen/native/Linear.cpp
Comment (Darktex) [inline aten/src/ATen/native/Linear.cpp:313]: I think this check is breaking Opacus

Comment (mruberry) [reply aten/src/ATen/native/Linear.cpp:313]: Hey @Darktex, could you elaborate?

Comment (heitorschueroff) [reply aten/src/ATen/native/Linear.cpp:313]: It could be, I didn't notice this at first but the previous version does not enforce this rule so this check is BC breaking. @mruberry I can send a quick patch to remove this rule and allow ellipsis to be reduced or I can add a BC breaking note. (numpy has this enforcement but it doesn't seem necessary).

**Comment by heitorschueroff** [comment]
Closing because of ghstack issue

**Pull request closed without merging**
